// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: middleware/v1/validator/validator.proto

package validatorv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Validator with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Validator) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Validator with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ValidatorMultiError, or nil
// if none found.
func (m *Validator) ValidateAll() error {
	return m.validate(true)
}

func (m *Validator) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Enabled

	if val := m.GetVersion(); val <= 0 || val >= 3 {
		err := ValidatorValidationError{
			field:  "Version",
			reason: "value must be inside range (0, 3)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for FailFast

	if len(errors) > 0 {
		return ValidatorMultiError(errors)
	}

	return nil
}

// ValidatorMultiError is an error wrapping multiple validation errors returned
// by Validator.ValidateAll() if the designated constraints aren't met.
type ValidatorMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ValidatorMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ValidatorMultiError) AllErrors() []error { return m }

// ValidatorValidationError is the validation error returned by
// Validator.Validate if the designated constraints aren't met.
type ValidatorValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ValidatorValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ValidatorValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ValidatorValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ValidatorValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ValidatorValidationError) ErrorName() string { return "ValidatorValidationError" }

// Error satisfies the builtin error interface
func (e ValidatorValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sValidator.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ValidatorValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ValidatorValidationError{}
